⭐️ manual client creation (with proxy)

joelddiaz opened the pull request:
Keep DefaultHttpClient() the way it is, but introduce a new HttpClient() that will allow specifying non-default values for the client.

Specifically, allow passing in a proxy string that will used for connections.

chris-rock commented on httpclient.go:
```diff
@@ -25,9 +32,60 @@ func DefaultHttpClient() *http.Client {
+	Proxy string
+}
+
+func HttpClient(opts *HttpClientOpts) (*http.Client, error) {
```
should this be `NewHttpClient`?

joelddiaz replied:
I followed the naming scheme from the existing `DefaultHttpClient()` https://github.com/mondoohq/ranger-rpc/blob/main/httpclient.go#L15 , but this is easy enough to change. Just let me know if I should only change this to NewHttpClient() or also change to NewDefaultHttpClient(), or some combination of these options.

chris-rock commented on httpclient.go:
```diff
@@ -25,9 +32,60 @@ func DefaultHttpClient() *http.Client {
+			return nil, err
+		}
+
+		proxyFn = func(*http.Request) (*url.URL, error) {
```
why not use `https://pkg.go.dev/net/http#ProxyURL`

joelddiaz replied:
Because I didn't know about this :) .

chris-rock commented on httpclient.go:
```diff
@@ -12,9 +15,13 @@ var (
 )
 
-func DefaultHttpClient() *http.Client {
+func newHttpTransport(proxy string) (*http.Transport, error) {
```
why not use:

```
func newHttpTransport(proxy *net.Url) *http.Client {
    proxyFn := http.ProxyFromEnvironment
    if proxy != nil {
      proxyFn = http.ProxyUrl(proxy)
    }
    
    tr := &http.Transport{
		Proxy: http.ProxyFromEnvironment,
		Proxy: proxyFn,
		...
  }
}
```

chris-rock commented on httpclient.go:
```diff
@@ -25,9 +32,60 @@ func DefaultHttpClient() *http.Client {
+// newProxyFunc will use a standard ProxyFromEnvironment if no proxy
+// information is provided. Otherwise, parse the proxy string, and use
+// it as the proxy endpoint.
+func newProxyFunc(proxy string) (proxyFunc, error) {
```
I think we do not need this function, makes the code to complex

chris-rock reviewed (changes requested):
Thank you @joelddiaz I added some feedback

joelddiaz commented:
@chris-rock addressed PR feedback. Only outstanding item is the naming of the function https://github.com/mondoohq/ranger-rpc/pull/25#discussion_r1140653174

chris-rock reviewed (approved):
Thank you @joelddiaz
